pkg/metadata/main.go

```diff
@@ -139,70 +139,80 @@ func main() {
 		// Userdata is no JSON, presumably...
 		log.Printf("Could not unmarshall userdata: %s", err)
 		// This is not an error
 		return nil
 	}
-	cm, ok := fd.(map[string]interface{})
-	if !ok {
-		log.Printf("Could convert JSON to desired format: %s", fd)
-		return nil
-	}
-	for d, val := range cm {
-		dir := path.Join(ConfigPath, d)
-		if err := os.MkdirAll(dir, 0755); err != nil {
-			log.Printf("Failed to create %s: %s", dir, err)
-			continue
-		}
-		files, ok := val.(map[string]interface{})
-		if !ok {
-			log.Printf("Could convert JSON for files: %s", val)
-			continue
+
+	var queue []entry
+	var current entry
+	queue = enqueue(queue, root, basePath)
+	for len(queue) != 0 {
+		current, queue = queue[0], queue[1:]
+		if current.Content != nil {
+			writeFile(current)
+		} else {
+			writeDirectory(current)
+			queue = enqueue(queue, current.Entries, current.path)
 		}
-		for f, i := range files {
-			p := uint64(0644)
-			var c string
-
-			switch fi := i.(type) {
-			case map[string]interface{}:
-				if _, ok := fi["perm"]; !ok {
-					log.Printf("No permission provided %s", f)
-					continue
-				}
-				if _, ok := fi["content"]; !ok {
-					log.Printf("No content provided %s", f)
-					continue
-				}
-				c = fi["content"].(string)
-				if p, err = strconv.ParseUint(fi["perm"].(string), 8, 32); err != nil {
-					log.Printf("Failed to parse permission %s: %s", fi, err)
-					continue
-				}
-			case string:
-				c = fi
-			default:
-				log.Printf("Couldn't convert JSON for items: %s", i)
-				continue
-			}
+	}
+	return nil
+}
 
-			if err := ioutil.WriteFile(path.Join(dir, f), []byte(c), os.FileMode(p)); err != nil {
-				log.Printf("Failed to write %s/%s: %s", dir, f, err)
-				continue
+func enqueue(queue []entry, entries map[string]entry, parent string) []entry {
+	for dir, entry := range entries {
+		entry.path = path.Join(parent, dir)
+		queue = append(queue, entry)
+	}
+	return queue
+}
 
-			}
+func writeDirectory(dir entry) {
+	filemode, err := parseFileMode(dir.Perm, 0755)
+	if err != nil {
+		log.Printf("Failed to parse permission %s: %s", dir, err)
+	}
+	if err := os.MkdirAll(dir.path, filemode); err != nil {
+		log.Printf("Failed to create %s: %s", dir.path, err)
+	}
+}
+
+func writeFile(file entry) {
+	filemode, err := parseFileMode(file.Perm, 0644)
+	if err != nil {
+		log.Printf("Failed to parse permission %s: %s", file, err)
+	}
+	if err := ioutil.WriteFile(file.path, []byte(*file.Content), filemode); err != nil {
+		log.Printf("Failed to write %s: %s", file.path, err)
+	}
+}
+
+func parseFileMode(input string, defaultMode os.FileMode) (os.FileMode, error) {
+	if input != "" {
+		perm, err := strconv.ParseUint(input, 8, 32)
+		if err != nil {
+			return 0, err
 		}
+		return os.FileMode(perm), nil
 	}
+	return defaultMode, nil
+}
 
-	return nil
+type entry struct {
+	path string
+
+	Perm    string           `json:"perm"`
+	Content *string          `json:"content"`
+	Entries map[string]entry `json:"entries"`
```

Comment:
I wonder if these should all be `,omitempty`?

Reply:
I thought you had said you wanted to be able to write the json out with this too? Perhaps I misremember, but it seems like a reasonable usecase anyway and perhaps would even justify making it `type Entry` so it can be imported?

Reply:
Yes, I will do that also. You're totally right. `Entry` instead of `entry`.